File: bullmq_worker/storage/minio_client.py
import os
from typing import Tuple
from urllib.parse import urlparse, unquote
from datetime import timedelta
from minio import Minio
from minio.error import S3Error

from bullmq_worker.utils.logging import logger
from bullmq_worker.storage.base_storage_client import BaseStorageClient


class MinIOClient(BaseStorageClient):
    """MinIO客户端"""

    def __init__(self, endpoint: str, access_key: str, secret_key: str, region: str, bucket_name: str):
        super().__init__(endpoint, access_key, secret_key, region, bucket_name)
        
        # 检测并设置secure模式和endpoint
        self.endpoint, self.secure = self._detect_secure_mode(endpoint)

        # 创建MinIO客户端
        self.client = Minio(
            endpoint=self.endpoint,
            access_key=access_key,
            secret_key=secret_key,
            secure=self.secure,
            region=region,
        )

        protocol = "HTTPS" if self.secure else "HTTP"
        logger.info(f"MinIO客户端初始化成功 - 服务器: {protocol}://{self.endpoint}, 桶: {bucket_name}")

        # 确保桶存在
        self.ensure_bucket_exists(bucket_name)

    def _detect_secure_mode(self, endpoint: str) -> Tuple[str, bool]:
        """检测并设置secure模式
        
        Args:
            endpoint: 原始endpoint
            
        Returns:
            tuple: (处理后的endpoint, secure模式)
        """
        
        # 自动检测secure模式
        if endpoint.startswith('https://'):
            return endpoint.replace('https://', ''), True
        elif endpoint.startswith('http://'):
            return endpoint.replace('http://', ''), False
        # 如果是localhost或127.0.0.1，默认使用HTTP
        elif 'localhost' in endpoint or '127.0.0.1' in endpoint:
            return endpoint, False
        else:
            # 其他情况默认使用HTTPS
            return endpoint, True

    def ensure_bucket_exists(self, bucket: str):
        """确保指定桶存在"""
        if not self.client.bucket_exists(bucket):
            try:
                self.client.make_bucket(bucket, location=self.region)
                logger.info(f"创建桶: {bucket}")
            except Exception as e:
                logger.warning(f"创建桶失败 {bucket}: {e}")
                raise RuntimeError(f"创建桶失败 {bucket}: {e}")
        else:
            logger.info(f"桶 {bucket} 存在")

    def download_file(self, object_key: str, local_path: str):
        """从MinIO下载文件"""
        try:
            # 创建本地目录
            os.makedirs(os.path.dirname(local_path), exist_ok=True)

            # 下载文件
            self.client.fget_object(
                bucket_name=self.bucket_name,
                object_name=object_key,
                file_path=local_path,
            )

            logger.info(f"下载成功: {self.bucket_name}/{object_key} -> {local_path}")

        except S3Error as e:
            if e.code == "NoSuchKey":
                logger.error(f"文件不存在: {self.bucket_name}/{object_key}")
                raise RuntimeError(f"文件不存在: {self.bucket_name}/{object_key}")
            else:
                logger.error(f"MinIO下载失败: {e}")
                raise RuntimeError(f"MinIO下载失败: {e}")
        except Exception as e:
            logger.error(f"下载文件时发生错误: {e}")
            raise RuntimeError(f"下载文件时发生错误: {e}")
    
    def upload_file(self, object_key: str, local_path: str, **kwargs) -> str:
        """上传文件到MinIO"""
        try:
            # 检查本地文件是否存在
            if not os.path.exists(local_path):
                raise FileNotFoundError(f"本地文件不存在: {local_path}")

            # 上传文件
            self.client.fput_object(
                bucket_name=self.bucket_name,
                object_name=object_key,
                file_path=local_path,
            )
            logger.info(f"上传成功: {local_path} -> {self.bucket_name}/{object_key}")

            # 生成预签名URL
            output_url = self.build_url(self.bucket_name, object_key)
            logger.info(f"输出URL: {output_url}")

            return output_url

        except S3Error as e:
            logger.error(f"MinIO上传失败: {e}")
            raise RuntimeError(f"MinIO上传失败: {e}")
        except Exception as e:
            logger.error(f"上传文件时发生错误: {e}")
            raise RuntimeError(f"上传文件时发生错误: {e}")

    def build_url(self, bucket: str, object_key: str, signer_url_expire_hours: int = 24) -> str:
        """构建MinIO URL - 生成预签名URL"""
        try:
            # 生成预签名URL
            expire_seconds = timedelta(hours=signer_url_expire_hours)
            signed_url = self.client.presigned_get_object(
                bucket_name=bucket,
                object_name=object_key,
                expires=expire_seconds,
            )

            logger.info(f"生成预签名URL: {signed_url}，有效期: {signer_url_expire_hours}小时")
            return signed_url
        
        except Exception as e:
            logger.error(f"生成预签名URL失败: {e}")
            raise RuntimeError(f"生成预签名URL失败: {e}")

    def parse_url(self, url: str) -> Tuple[str, str]:
        """解析MinIO URL"""
        parsed = urlparse(url)
        hostname = parsed.hostname or ""
        port = parsed.port

        # 构建完整的endpoint用于比较
        endpoint_to_check = f"{hostname}:{port}" if port else hostname
        
        # MinIO URL格式: http://127.0.0.1:9000/bucket/path/file
        # 检查hostname是否匹配（忽略协议）
        if (hostname in self.endpoint or 
            endpoint_to_check in self.endpoint or 
            self.endpoint.replace('http://', '').replace('https://', '') == endpoint_to_check):
            
            path_parts = parsed.path.strip("/").split("/", 1)
            if len(path_parts) >= 2:
                bucket = path_parts[0]
                object_key = unquote(path_parts[1])
                return bucket, object_key
            elif len(path_parts) == 1:
                # 如果只有一个路径部分，可能是bucket.endpoint格式
                parts = hostname.split(".")
                if len(parts) >= 2:
                    bucket = parts[0]
                    object_key = unquote(path_parts[0])
                    return bucket, object_key

        raise ValueError(f"无法解析MinIO URL: {url}")


def test_client():
    from dotenv import load_dotenv, find_dotenv
    load_dotenv(find_dotenv())

    client = MinIOClient(
        endpoint=os.getenv("S3_ENDPOINT"),
        access_key=os.getenv("S3_ACCESS_KEY"),
        secret_key=os.getenv("S3_SECRET_KEY"),
        bucket_name=os.getenv("S3_BUCKET"),
        region=os.getenv("S3_REGION"),
    )

    # 上传文件
    local_path = "assets/demo1_audio.wav"
    url = client.upload_file(
        object_key="wjw_demo1_audio.wav",
        local_path=local_path
    )

    # 下载文件
    bucket, object_key = client.parse_url(url)
    client.download_file(
        object_key=object_key,
        local_path=f"downloads/{object_key}"
    )


if __name__ == "__main__":
    test_client()
    

File: bullmq_worker/src/storage/minio-client.ts
import * as Minio from 'minio';
import * as fs from 'fs';
import * as path from 'path';
import { URL } from 'url';
import { BaseStorageClient, StorageConfig } from './base-client';
import logger from '../utils/logger';

/**
 * MinIO 存储客户端实现
 */
export class MinIOClient extends BaseStorageClient {
  private client: Minio.Client;
  private secure: boolean;
  
  constructor(config: StorageConfig) {
    super(config);
    
    // 检测并设置安全模式
    const { endpoint, secure } = this.detectSecureMode(config.endpoint);
    this.secure = secure;
    
    // 提取端口和主机名
    const { hostname, port } = this.parseEndpoint(endpoint);
    this.endpoint = hostname;
    
    // 创建 MinIO 客户端
    this.client = new Minio.Client(
      {
        endPoint: hostname,
        ...(port && { port: port }),
        useSSL: this.secure,
        accessKey: this.accessKey,
        secretKey: this.secretKey,
        ...(this.region && { region: this.region }),
      }
    );
    
    const protocol = this.secure ? 'HTTPS' : 'HTTP';
    const displayEndpoint = port ? `${hostname}:${port}` : hostname;
    logger.info(`✅ MinIO客户端初始化成功 - 服务器: ${protocol}://${displayEndpoint}, 桶: ${this.bucketName}`);
    
    // 检查桶是否存在
    this.ensureBucketExists().catch(error => {
      logger.error(error.message || error);
    });
  }

  /**
   * 检测并设置secure模式
   */
  private detectSecureMode(endpoint: string): { endpoint: string; secure: boolean } {
    // 如果有协议前缀，根据协议确定安全模式
    if (endpoint.startsWith('https://')) {
      return { endpoint: endpoint.replace('https://', ''), secure: true };
    } else if (endpoint.startsWith('http://')) {
      return { endpoint: endpoint.replace('http://', ''), secure: false };
    }
    
    // 没有协议前缀的情况
    // localhost 和 127.0.0.1 默认使用 HTTP
    if (endpoint.includes('localhost') || endpoint.includes('127.0.0.1')) {
      return { endpoint, secure: false };
    }
    
    // 其他情况默认使用 HTTP（适用于内网和自建服务）
    return { endpoint, secure: false };
  }
  
  /**
   * 解析端点，提取主机名和端口（不会修改实例属性）
   */
  private parseEndpoint(endpoint: string): { hostname: string; port?: number } {
    const parts = endpoint.split(':');
    if (parts.length > 1) {
      const port = parseInt(parts[parts.length - 1], 10);
      if (!isNaN(port)) {
        const hostname = parts.slice(0, -1).join(':');
        return { hostname, port };
      }
    }
    return { hostname: endpoint };
  }
  
  /**
   * 确保桶存在
   */
  protected async ensureBucketExists(): Promise<void> {
    try {
      const exists = await this.client.bucketExists(this.bucketName);
      if (!exists) {
        await this.client.makeBucket(this.bucketName, this.region);
        logger.info(`🔧 创建桶: ${this.bucketName}`);
      } else {
        logger.info(`📦 桶 ${this.bucketName} 存在`);
      }
    } catch (error) {
      throw new Error(`检查桶状态失败: ${error}`);
    }
  }
  
  /**
   * 从MinIO下载文件
   */
  async downloadFile(objectKey: string, localPath: string): Promise<void> {
    try {
      // 创建本地目录
      const dir = path.dirname(localPath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      
      // 下载文件
      await this.client.fGetObject(this.bucketName, objectKey, localPath);
      logger.info(`✅ 下载成功: ${this.bucketName}/${objectKey} -> ${localPath}`);
    } catch (error: any) {
      if (error.code === 'NoSuchKey') {
        logger.error(`❌ 文件不存在: ${this.bucketName}/${objectKey}`);
        throw new Error(`文件不存在: ${this.bucketName}/${objectKey}`);
      } else {
        logger.error(`❌ MinIO下载失败: ${error}`);
        throw new Error(`MinIO下载失败: ${error}`);
      }
    }
  }
  
  /**
   * 上传文件到MinIO
   */
  async uploadFile(objectKey: string, localPath: string): Promise<string> {
    try {
      // 检查本地文件是否存在
      if (!fs.existsSync(localPath)) {
        throw new Error(`❌ 本地文件不存在: ${localPath}`);
      }
      
      // 上传文件
      await this.client.fPutObject(this.bucketName, objectKey, localPath);
      logger.info(`✅ 上传成功: ${localPath} -> ${this.bucketName}/${objectKey}`);
      
      // 生成预签名URL
      const outputUrl = await this.buildUrl(objectKey);
      return outputUrl;
    } catch (error) {
      logger.error(`❌ MinIO 上传失败: ${error}`);
      throw new Error(`MinIO 上传失败: ${error}`);
    }
  }
  
  /**
   * 生成预签名URL
   */
  async buildUrl(objectKey: string, expiresInHours: number = 24): Promise<string> {
    try {
      const expireSeconds = expiresInHours * 3600;
      const signedUrl = await this.client.presignedGetObject(
        this.bucketName,
        objectKey,
        expireSeconds
      );
      
      logger.info(`✅ 生成预签名URL: ${signedUrl}, 有效期: ${expiresInHours} 小时`);
      return signedUrl;
    } catch (error) {
      logger.error(`❌ 生成预签名URL失败: ${error}`);
      throw new Error(`生成预签名URL失败: ${error}`);
    }
  }
  
  /**
   * 解析 MinIO URL
   */
  parseUrl(url: string): { bucket: string; objectKey: string } {
    try {
      const parsed = new URL(url);
      const hostname = parsed.hostname || '';
      const port = parsed.port;
      
      // 构建完整的endpoint用于比较
      const endpointToCheck = port ? `${hostname}:${port}` : hostname;
      
      // # MinIO URL格式: http://127.0.0.1:9000/bucket/path/file
      // 检查hostname是否匹配（忽略协议）
      const cleanEndpoint = this.endpoint.replace('http://', '').replace('https://', '');
      
      if (hostname === cleanEndpoint || endpointToCheck === cleanEndpoint || 
          hostname.includes(cleanEndpoint) || cleanEndpoint.includes(hostname)) {
        
        const pathParts = parsed.pathname.substring(1).split('/', 2);
        if (pathParts.length >= 2) {
          const bucket = pathParts[0];
          const objectKey = decodeURIComponent(pathParts[1]);
          return { bucket, objectKey };
        } else if (pathParts.length === 1) {
          // 如果只有一个路径部分，可能是bucket.endpoint格式
          const hostParts = hostname.split('.');
          if (hostParts.length >= 2) {
            const bucket = hostParts[0];
            const objectKey = decodeURIComponent(pathParts[0]);
            return { bucket, objectKey };
          }
        }
      }
      
      throw new Error(`无法解析 MinIO URL: ${url}`);
    } catch (error) {
      logger.error(`解析URL失败: ${error}`);
      throw new Error(`解析URL失败: ${error}`);
    }
  }
}
